File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root: TreeNode) -> int:
        def dfs(node):
            if node is None:
                return {"sum": 0, "node_count": 0, "count": 0}

            left = dfs(node.left)
            right = dfs(node.right)

            total_sum = left["sum"] + right["sum"] + node.val
            total_node_count = left["node_count"] + right["node_count"] + 1

            average = total_sum // total_node_count

            if node.val == average:
                count = left["count"] + right["count"] + 1
            else:
                count = left["count"] + right["count"]

            return {"sum": total_sum, "node_count": total_node_count, "count": count}

        return dfs(root)["count"]

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var averageOfSubtree = function (root) {
    const dfs = function (node) {
        if (node === null) {
            return { sum: 0, count: 0, nodeCount: 0 };
        }

        let { sum: leftSum, nodeCount: leftNodeCount, count: leftCount } = dfs(node.left);
        let { sum: rightSum, nodeCount: rightNodeCount, count: rightCount } = dfs(node.right);

        let totalSum = leftSum + rightSum + node.val;
        let totalNodeCount = leftNodeCount + rightNodeCount + 1;

        let average = Math.floor(totalSum / totalNodeCount);

        if (average === node.val) {
            return { sum: totalSum, nodeCount: totalNodeCount, count: leftCount + rightCount + 1 }
        } else {
            return { sum: totalSum, nodeCount: totalNodeCount, count: leftCount + rightCount }
        }
    }

    return dfs(root).count;
};